psi4/src/psi4/libfunctional/superfunctional.cc

```diff
@@ -381,6 +386,51 @@ void SuperFunctional::add_c_functional(std::shared_ptr<Functional> fun) {
     can_edit();
     c_functionals_.push_back(fun);
 }
+void SuperFunctional::copy_density_tolerance(double cut) {
+    can_edit();
+    density_tolerance_ = cut;
+}
+void SuperFunctional::set_density_tolerance(double cut) {
+    // modify both LibXCFunctional and Functional
+    density_tolerance_ = cut;
+    if (libxc_xc_func_) {
+        for (int Q = 0; Q < c_functionals_.size(); Q++) {
+            c_functionals_[Q]->set_density_cutoff(density_tolerance_);
+            dynamic_cast<LibXCFunctional*>(c_functionals_[Q].get())->set_density_cutoff(density_tolerance_);
+        }
+    } else {
+        for (int Q = 0; Q < c_functionals_.size(); Q++) {
+            c_functionals_[Q]->set_density_cutoff(density_tolerance_);
+            dynamic_cast<LibXCFunctional*>(c_functionals_[Q].get())->set_density_cutoff(density_tolerance_);
+        };
+        for (int Q = 0; Q < x_functionals_.size(); Q++) {
+            x_functionals_[Q]->set_density_cutoff(density_tolerance_);
+            dynamic_cast<LibXCFunctional*>(x_functionals_[Q].get())->set_density_cutoff(density_tolerance_);
+        };
+    };
+}
+void SuperFunctional::print_density_threshold(std::string out, int level) const {
+    if (level < 1) return;
+    std::shared_ptr<psi::PsiOutStream> printer = (out == "outfile" ? outfile : std::make_shared<PsiOutStream>(out));
+    printer->Printf("   => LibXC Density Thresholds  <==\n\n");
+    double val = 0.0;
+    if (libxc_xc_func_) {
+        for (int Q = 0; Q < c_functionals_.size(); Q++) {
+            val = dynamic_cast<LibXCFunctional*>(c_functionals_[Q].get())->query_density_cutoff();
```

Comment:
Could we have getters/setters for the Functional class which raise a not implemented error and then put the specific implementation in LibXCFunctional so that we do not need to dynamic cast here?

Reply:
Do you mean just here or also in the function above?

Reply:
OK, today I learned about what a virtual function is and what those `override` commands do. Very helpful. Will remove all those dynamic casts I introduced.